Add support for TLS/SSL enabled clusters

damienclaveau opened the pull request:
Here is a patch to support SSL enabled Hadoop services.
Tested in our environment with Yarn and HDFS 2.7.1, works fine :-)

You will need to add jvm_args="-Djavax.net.ssl.trustStore=path_to_truststore" in elephant.conf

There is still one error with the swebhdfs client in the SparkFSFetcher.scala class complaining about a SSL certificate problem. 
Seems to be related to a dependency brought by sbt. I have not that error in a simple unit test project built with Maven. 

Any review will be welcome.

Commit:
HADOOP-8320: Add detailed suggestions in Dr. Elephant help page

Commit:
HADOOP-8320: Add detailed suggestions in Dr. Elephant help page

Commit:
HADOOP-8856: Fix a broken Dr.Elephant test case for reducer time heuristic

Commit:
HADOOP-8859: Add ideal task time suggestion in Dr. Elephant 's reducer time help page

Commit:
HADOOP-8846: Release Dr. Elephant v0.6.5

Commit:
HADOOP-8929: Update Dr.Elephant version number to 0.6.6-SNAPSHOT

Commit:
HADOOP-8941: Further improve search performance for Dr. Elephant

Commit:
HADOOP-9685: Dr. Elephant fails to fetch task counter data on hadoop2

Commit:
HADOOP-9714: Add Pluggable jobtype to Dr. Elephant

Commit:
HADOOP-9716: Dr. Elephant rest interface needs search endpoint

Commit:
HADOOP-9853: Release Dr. Elephant v0.6.6

Commit:
HADOOP-9886: Fix for making Dr. Elephant's pagination thread safe

Commit:
HADOOP-10289: Update Dr. Elephant version number to 0.6.7-SNAPSHOT

Commit:
HADOOP-10089: Dr. Elephant misses on type of Voldemort bnp job

Commit:
HADOOP-10301: Dr. Elephant occasionally misses a job

Commit:
HADOOP-9900: Dr. Elephant Mapper Input Size Help Page should suggest increasing mapred.min.split.size for too many mappers, NOT mapred.max.split.size

Commit:
HADOOP-10290: Release Dr. Elephant v0.6.7

Commit:
HADOOP-10314: Dr. Elephant should not mention deprecated properties in Help topics page

mcvsubbu committed:
[HADOOP-10405] Fix one-off error in getting cluster name from NN hostname

RB=468832
BUGS=HADOOP-10405
R=fli
A=fli,ahsu

timyitong committed:
OFFREL-234: Adding Spark log analysers into Dr. Elephant

timyitong committed:
Addressing naming problems, code cleaness and added more help pages

timyitong committed:
Renaming Mapreduce -> MapReduce, Classic -> Hadoop1, Yarn -> Hadoop2

timyitong committed:
Addressing major feedback from Fangshi: 1.bind application type with JobType; 2. use only application id all data holders and use job id while saving to DB; 3. Change configuration styles of heuristics, fetchers and jobtypes into separate files; 4. Make those configuration files providable via other locations so that we can seemlessly change configuration without releasing a new deployment zip. 5. Remove HadoopVersion classic, yarn, using 1 or 2 instead.

timyitong committed:
Modify Spark's dependency configuration, some more logging and a fix of SparkFSFetcher to avoid sharing the same filesystem object across different threads

timyitong committed:
A tmp commit trying to test Nertz Spades deployment

timyitong committed:
Addressing some renaming feedback, logging

Commit:
HADOOP-11768: Update Dr. Elephant version to v1.0.0-SNAPSHOT

timyitong committed:
Code cleaness and styles improvements

RB=488607
R=fli,mwagner
A=fli

Commit:
HDOOP-11456: Change Dr. Elephant's heuristic from MapperInputSize to MapperTime

Commit:
HADOOP-11876: Separate Dr. Elephant configs from deployable zip

Commit:
HADOOP-11922: Fix Dr. Elephant search by heuristic field

Commit:
HADOOP-11912: Release Dr. Elephant v1.0.0

Commit:
HADOOP-12013: Dr. Elephant has a corner case to analyze duplicate entries

Commit:
HADOOP-12247 - Update Dr. Elephant version to v1.0.1-SNAPSHOT

Commit:
HADOOP-11913: Fix Dr. Elephant broken tests

Commit:
HADOOP-12067: Dr. Elephant heuristic is too lenient to jobs with long running tasks

Commit:
HADOOP-12023: Dr. Elephagent new heuristics: memory util ratio

Commit:
HADOOP-12593: Release Dr. Elephant v1.0.1

Commit:
HADOOP-12632: Enable Dr. Elephant 's mapper spill heuristics

Commit:
HADOOP-12669: Dr. Elephant's mapper spill heuristic is too harsh

Commit:
HADOOP-12674: Release Dr. Elephant v1.0.2

Commit:
HADOOP-12603: Dr. Elephant has a failed test in Mapper Speed Heuristic

Commit:
HADOOP-12603: Update Mapper Spill Heuristic help page

Commit:
HADOOP-12063: fix a potential devide by 0 exception in mapper spill heuristic

Commit:
HADOOP-12688: Dr. Elephant needs task time info in Memory Heuristic

Commit:
HADOOP-12709: Dr. Elephant has a corner case for flagging 0 tasks in Memory Heuristic

timyitong committed:
LIHADOOP-13019, LIHADOOP-13020: Upgrade to Spark 1.4, and add event log limit for Spark log fetcher

RB=526958
BUG=
G=superfriends-reviewers
R=fli
A=fli

Commit:
LIHADOOP-13018: Dr. Elephant 's Spill heuristic should be lenient to small number of tasks

Commit:
LIHADOOP-13094: Fix Dr. Elephant build on Jenkins after spark1.4 upgrade

Commit:
LIHADOOP-13016: Release Dr. Elephant v1.0.3

Victsm committed:
LIHADOOP-13747 Update Spark driver name to be consistent with Spark 1.4.

timyitong committed:
Adjust StageRuntime heuristics so that it's the real duration time instead of total aggregated executor time

RB=549801

G=superfriends-reviewers
R=mshen
A=mshen

akshayrai committed:
LIHADOOP-8114 [Opensource Dr. E] Apache License and LinkedIn Copyright Notices

RB=551175

G=superfriends-reviewers
R=abain,ahsu,annag,csteinba,fli,mwagner,rratti
A=csteinba

Commit:
LIHADOOP-5177 Update documentation of Dr. Elephant

RB=554506

G=superfriends-reviewers
R=annag,fli,rratti,mwagner
A=fli,vigankid

Commit:
LIHADOOP-13953 The start script should take arguments to enable evolutions

RB=554498

BUGS=LIHADOOP-13953
G=superfriends-reviewers
R=fli, rratti, annag
A=fli

Commit:
LIHADOOP-13711: Dr. Elephant should add virtual memory info in its memory heuristics

Commit:
LIHADOOP-12802 Occasional Concurrent Modification Exception during Dr. Elephant start

RB=563017

G=superfriends-reviewers
R=rratti,fli
A=rratti

Commit:
Minor formatting change in Readme License

Commit:
LIHADOOP-14305: Dr. Elephant goes OOM while displaying other executions of a job

Commit:
LIHADOOP-14319: Dr. Elephant's should append not overwrite log during restart

Commit:
LIHADOOP-14321: Release Dr. Elephant v1.0.4

Commit:
LIHADOOP-14322 Dr. Elephant feature request. The ability to compare result for two flow executions

RB=576207
BUGS=LIHADOOP-14322

G=superfriends-reviewers
R=fli,rratti
A=fli

timyitong committed:
LIHADOOP-14722: Adjust some Spark heuristic parameters of Dr Elephant

RB=577602
G=superfriends-reviewers
R=fli,mshen
A=mshen,fli

Commit:
LIHADOOP-14688 Adding a timestamp to Dr. Elephant search result page

RB=578468

G=superfriends-reviewers
R=fli,jdintruf
A=fli

Commit:
LIHADOOP-15080 Clean up and organize the views in Dr. Elephant

G=superfriends-reviewers
R=annag,fli,rratti
A=fli,annag

Commit:
LIHADOOP-13819: Dr. Elephant heuristic request. GC efficiency

Commit:
LIHADOOP-15134: Dr. Elephant should provide exception stacktrace from failed jobs

Conflicts:
	build.sbt

Commit:
LIHADOOP-14342: Dr. Elephant's memory heuristic has a corner case to throw NumberFormatException

Commit:
LIHADOOP-15136: Release Dr. Elephant v1.0.5

Commit:
LIHADOOP-15252: Update Dr. Elephant version to v1.0.6-SNAPSHOT

akshayrai committed:
LIHADOOP-15372 Update the Heuristic Conf for helpGC with the proper package name

G=superfriends-reviewers
R=fli,rratti
A=rratti

akshayrai committed:
LIHADOOP-15377 Job execution URLs on Dr. Elephant's Flow Execution page is incorrect

RB=598350

G=superfriends-reviewers
R=ahsu,fli
A=ahsu

Commit:
LIHADOOP-15268: Task sampling in Dr. Elephant

Commit:
LIHADOOP-16125 Other executions of a job should list the recent jobs from history firBt

RB=626046

G=superfriends-reviewers
R=annag,fli,mwagner,rratti,viramach
A=fli

Commit:
LIHADOOP-15217 Implement comparison of historical flows/jobs in Dr. Elephant

RB=629171

G=superfriends-reviewers
R=annag,fli,mwagner,rratti,viramach
A=annag,fli,rratti

Commit:
LIHADOOP-16102 Discuss and remove Hadoop 1 support from Dr. Elephant

RB=634544

G=superfriends-reviewrs
R=annag,fli,rratti,viramach
A=rratti

Commit:
LIHADOOP-16576 Make severity threshold values configurable for the heuristics.

RB=635070

G=superfriends-reviewers
R=annag,fli,rratti,shanm,viramach
A=fli

Commit:
LIHADOOP-9004 Support Hadoop 2 properties in Dr. Elephant

RB=635540

G=superfriends-reviewers
R=annag,fli,rratti,viramach
A=annag,fli

Commit:
LIHADOOP-16687 Update Dr. Elephant to use the new Hadoop API

RB=636803

G=superfriends-reviewers
R=annag,fli,rratti,viramach
A=rratti

akshayrai committed:
LIHADOOP-16587 Dr. Elephant should provide search by start time

RB=646592

G=superfriends-reviewers
R=annag,fli,rratti,shanm,viramach
A=fli

akshayrai committed:
LIHADOOP-17131 Prepare Readme for Dr. Elephant

RB=650949

G=superfriends-reviewers
R=annag,fli,rratti,shanm,viramach
A=viramach

akshayrai committed:
LIHADOOP-17485: Refactor and cleanup Dr. Elephant code

RB=660706

G=superfriends-reviewers
R=annag,fli,rratti,shanm,viramach
A=annag

akshayrai committed:
LIHADOOP-17502: Improve Dr. Elephant's history graph and tooltips

RB=661471

G=superfriends-reviewers
R=annag,fli,rratti,shanm,viramach
A=fli

akshayrai committed:
LIHADOOP-17538: Code changes to support new schema

RB=666697

G=superfriends-reviewers
R=annag,fli,rratti,shanm,viramach
A=fli

akshayrai committed:
LIHADOOP-16100: Dr. Elephant should show timestamp specific to timezones or at least UTC.

RB=661685

G=superfriends-reviewers
R=annag,fli,rratti,shanm,viramach
A=annag,fli

akshayrai committed:
LIHADOOP-17669: Update Notice and copyright year on Dr. Elephant

RB=668782

G=superfriends-reviewers
R=annag,fli,shanm,viramach
A=fli,shanm

akshayrai committed:
LIHADOOP-17537: Replace bootstrap-datepicker library with a copy from github

RB=668783

G=superfriends-reviewers
R=annag,fli,shanm,viramach
A=fli

akshayrai committed:
LIHADOOP-17715: Remove Publish Metrics Logic from Dr. Elephant

RB=668785

G=superfriends-reviewers
R=annag,fli,shanm,viramach
A=fli,shanm

akshayrai committed:
LIHADOOP-17783: Update mysql-connector-java version to 5.1.36

akshayrai committed:
LIHADOOP-15398: Release Dr. Elephant v2.0.0

akshayrai committed:
LIHADOOP-15398: Release Dr. Elephant v2.0.1 (The jenkins build failed with v2.0.0 due to a corrupt build)

akshayrai committed:
LIHADOOP-17890: Set default fields when spark fails to get the job conf

RB=671778

G=superfriends-reviewers
R=annag,fli,shanm
A=annag

akshayrai committed:
LIHADOOP-16103 Check and Separate Linkedin Specific Code from Dr. Elephant

RB=632808

G=superfriends-reviewers
R=annag,fli,mwagner,rratti,shanm,viramach
A=rratti,shanm

akshayrai committed:
LIHADOOP-16101: Review the evolution scripts in Dr. Elephant

RB=672888

G=superfriends-reviewers
R=annag,fli,shanm,viramach
A=shanm

akshayrai committed:
LIHADOOP-17996: Fix potential issues in Dr. Elephant v2.0.1

RB=676126

G=superfriends-reviewers
R=annag,fli,shanm,viramch
A=fli

akshayrai committed:
LIHADOOP-18028: Dr. Elephant 's spark parser on Nertz is not working

RB=677178

G=superfriends-reviewers
R=annag,fli,shanm,viramach
A=shanm

akshayrai committed:
LIHADOOP-18056: Release Dr. Elephant v2.0.2

akshayrai committed:
LIHADOOP-18124: Complete Open sourcing of Dr. Elephant

RB=681089

G=superfriends-reviewers
R=annag,fli,shanm,viramach
A=annag,shanm

akshayrai committed:
LIHADOOP-16126: Easy integration of new schedulers to Dr. Elephant

RB=676088

G=superfriends-reviewers
R=annag,fli,shanm,viramach
A=annag,shanm

akshayrai committed:
LIHADOOP-18232: Include a resolver template file in the configuration

akshayrai committed:
Merge pull request #1 from akshayrai/master

LIHADOOP-18232: Include a resolver template file in the configuration

akshayrai committed:
LIHADOOP-18234: Add wiki documentation images to code base

akshayrai committed:
Merge pull request #2 from akshayrai/addImages

LIHADOOP-18234: Add wiki documentation images to code base

akshayrai committed:
LIHADOOP-18434: Add index on start time and allow querying by start time in the rest api

akshayrai committed:
Merge pull request #3 from akshayrai/revertToBigint

LIHADOOP-18434: Add index on start time and allow querying by start t…

akshayrai committed:
LIHADOOP-18918: Dr. Elephant should support all play versions

akshayrai committed:
Merge pull request #4 from akshayrai/Support

LIHADOOP-18918: Dr. Elephant should support all play versions

akshayrai committed:
LIHADOOP-18919: Investigate and fix why Dr. Elephant configurations are not included in classpath

akshayrai committed:
Merge pull request #5 from akshayrai/confDirFix

LIHADOOP-18919: Investigate and fix why Dr. Elephant configurations a…

akshayrai committed:
LIHADOOP-18977: Dr. Elephant doesn't compile with spark 1.5.0+

akshayrai committed:
Merge pull request #9 from akshayrai/sparkTypeMismatch

LIHADOOP-18977: Dr. Elephant doesn't compile with spark 1.5.0+

Commit:
LIHADOOP-18925: Expose additional metrics through Dr Elephant ( Job Execution time, Job wait time, resource usage, wasted resources )

Commit:
Get the active resource manager URL when HA is enabled

Commit:
fix start script to redirect output to a file instead of /dev/null

Commit:
addressed code review comments on RM HA support PR

akshayrai committed:
Merge pull request #15 from krishnap/kp_fix_start_script

Fix start script to redirect output to a file instead of /dev/null

akshayrai committed:
Merge pull request #14 from krishnap/kp_fix_rm_issue

Get the active resource manager URL when HA is enabled

rajagopr committed:
LIHADOOP-19325: [DR-Elephant] Add integration tests for the rest API's in Dr. Elephant

* Added tests for the rest API endpoints.

The following endpoints are covered.
  1. /rest/job
  2. /rest/jobexec
  3. /rest/flowexec
  4. /rest/search
  5. /rest/compare
  6. /rest/flowgraphdata
  7. /rest/jobgraphdata

The tests are run against a test server which deploys a fake application connecting to an in-memory H2 DB.
The Global class is overridden in the tests so that we don't have to go through the regular application
start flow. Assertions are made on the json response from each of the above listed endpoints.

Results from 'play test' as below.

$>play test
...
...
[info] rest.RestAPITest
[info] + restFlowExecResult
[info] + testrestJobExecResult
[info] + testrestSearch
[info] + testrestJobGraphData
[info] + testrestAppResult
[info] + testrestFlowGraphData
[info] + testrestSearchWithUsernameAndJobType
[info] + testrestCompare
[info]
[info]
[info] Total for test rest.RestAPITest
[info] Finished in 0.001 seconds
[info] 8 tests, 0 failures, 0 errors
...
...
[info] Passed: Total 141, Failed 0, Errors 0, Passed 141
[success] Total time: 5 s, completed Apr 21, 2016 3:59:15 PM

* Incorporated the review comments.

1. Added License file to DBTestUtil.java
2. Explicitly turning on evolutions for test
3. Inserting multiple rows in a single INSERT statement
4. Fixed test method name to start with test

note: All tests passed in re-run.

akshayrai committed:
Added a link to the mailing list.

tglstory committed:
Adding commons codec to the dependency (#26)

tglstory committed:
Spark dependency resolution based on version (#25)

rajagopr committed:
LIHADOOP-19491: [DR-Elephant] Add code coverage to Dr. Elephant

* Javascript engine is returned as null when com.linkedin.drelephant.util.UtilsTest.testGetParam() is run
with code coverage. Turns out to be an sbt issue as noted here.
http://stackoverflow.com/questions/20168226/sbt-0-13-scriptengine-is-null-for-getenginebyname-javascript

* Adding jacoco code coverage to the project

Test results as noted below

$>play jacoco:cover
...
...
[info] Total for test com.linkedin.drelephant.mapreduce.heuristics.MapperTimeHeuristicTest
[info] Finished in 0.009 seconds
[info] 9 tests, 0 failures, 0 errors
[info] Passed: Total 141, Failed 0, Errors 0, Passed 141
[info]
[info] ------- Jacoco Coverage Report --------
[info]
[info] Lines: 47.85% (>= required 0.0%) covered, 1707 of 3273 missed, OK
[info] Instructions: 45.29% (>= required 0.0%) covered, 11864 of 21685 missed, OK
[info] Branches: 35.35% (>= required 0.0%) covered, 856 of 1324 missed, OK
[info] Methods: 44.66% (>= required 0.0%) covered, 456 of 824 missed, OK
[info] Complexity: 34.77% (>= required 0.0%) covered, 1013 of 1553 missed, OK
[info] Class: 52.67% (>= required 0.0%) covered, 62 of 131 missed, OK
[info] Check target/jacoco for detail report
[info]

Commit:
LIHADOOP-19495 Make sampling configurable in Dr elephant

akshayrai committed:
LIHADOOP-19609: Make Spark Event Log Directory configurable

Commit:
LIHADOOP-19610: Metrics should not be shown if runtime and resourceused are 0

Commit:
LIHADOOP-18924: Improve Job History to show runtime, waittime, resource used, resource wasted.

stiga-huang committed:
Fix an cast error in SparkDataCollection.scala (#37)

* Fix an cast error in SparkDataCollection.scala

In Spark 1.5 and later, org.apache.spark.ui.jobs.UIData. JobUIData#completedStageIndices is a mutable.HashSet instead of an OpenHashSet. To make our codes compatible for all Spark version, we need to override method addIntSetToJSet to accept OpenHashSet and mutable.HashSet both.

tglstory committed:
Separate configuration logic from RM address update logic (#38)

rajagopr committed:
LIHADOOP-19850: Test for AnalyticJob (#42)

* Test for AnalyticJob

Added test to perform analysis on an analytic job and compute results.
Dependencies in getAnalysis() method are mocked out using JMockit.

$>../tools/play "test-only com.linkedin.drelephant.analysis.AnalyticJobTest"
...
[info] com.linkedin.drelephant.analysis.AnalyticJobTest
[info] + testGetAnalysis
[info]
[info]
[info] Total for test com.linkedin.drelephant.analysis.AnalyticJobTest
[info] Finished in 0.01 seconds
[info] 1 tests, 0 failures, 0 errors
[info] Passed: Total 1, Failed 0, Errors 0, Passed 1
[success] Total time: 2 s, completed May 9, 2016 1:50:18 PM

Notes:
1. Jacoco and JMockit doesn't work well together. This test would have to be ignored if generating
code coverage with jacoco. Alternately, coverage can be generated with JMockit itself.

rajagopr committed:
LIHADOOP-20088: Add organization name to Dr. Elephant to generate ivy file (#46)

Commit:
LIHADOOP-20170: Write help pages for new metrics in Dr. Elephant

Commit:
LIHADOOP-19841: Enable Analyzing Spark jobs for Namenode HA

Commit:
LIHADOOP-19666: totalDelay is negative for zero tasks

Commit:
Merge remote-tracking branch 'origin/master' into AggregatedMetrics

Conflicts:
	test/common/TestConstants.java

Commit:
Minor fixup on AnalyticJobTest

Commit:
LIHADOOP-20350: Release dr-elephant v. 2.0.3

Commit:
LIHADOOP-20390: Disable AnalyticTest in Dr. Elephant

akshayrai committed:
LIHADOOP-20347: Fix Mapper Speed Heuristic's Disk Speed Severity Thresholds

Commit:
LIHADOOP-20392: NullPointerException on zero mappers or reducers

Commit:
Merge remote-tracking branch 'upstream/master' into AggregatedMetrics

akshayrai committed:
Added engineering blog link

rajagopr committed:
LIHADOOP-20476: Generating code coverage report in xml format too. This will be consumed by the product dashboard. (#65)

Commit:
LIHADOOP-20582: Dr Elephant shows all the analyzed jobs on the dashboard instead of last 24 hours

Commit:
LIHADOOP-20577: Dr Elephant shows more than 100% waittime for some jobs

Commit:
LIHADOOP-20626 Ignore slow start for considering reducer delay

Commit:
LIHADOOP-20629 Minor UI fixes for Dr. Elephant

* Increase space between the resource wasted icon and text
* Remove hours from the runtime
* Modify tooltip to for runtime
* Show runtime and waitime in hh:mm:ss in job history graph

Commit:
Merge branch 'master' of https://github.com/linkedin/dr-elephant into AggregatedMetrics

Commit:
LIHADOOP-20630: Release Dr.Elephant v. 2.0.4

rajagopr committed:
Removing some un-necessary logic. (#70)

chetnachaudhari committed:
Added support to search by queue-name (#40)

Added support to search by queue-name

paulbramsen committed:
Try alternate resource manager if one is down (#78)

paulbramsen committed:
Ignore and remove .DS_Store files (#80)

stiga-huang committed:
Add GeneralConf.xml to make number of executors configurable (#75)

paulbramsen committed:
Add spark log ext conf param (#79)

ljank committed:
Fix fallback to default `SPARK_LOG_EXT` value (#82)

paulbramsen committed:
Add Travis CI support (#86)

paulbramsen committed:
Add Airflow scheduler support to Dr. Elephant. Includes Tests. (#83)

paulbramsen committed:
Make default container mem configurable (#90)

rajagopr committed:
Web analytics for Dr. Elephant. (#94)

Made changes so that web analytics of choice can be configured into Dr. Elephant.
By default the feature will be disabled and can be turned on via configuration. Details on usage
is mentioned in the comments section in application.conf file.

paulbramsen committed:
Fix job history page tab highlighting (#98)

YannByron committed:
dr-elephant#95: copy conf files to zip package automatically (#96)

gitter-badger committed:
Add Gitter badge (#102)

paulbramsen committed:
Put build and gitter badges on the same line (#105)

Commit:
Update wiki images for aggregated metrics feature.

Commit:
Merge remote-tracking branch 'origin/AggregatedMetrics'

Conflicts:
	app/com/linkedin/drelephant/ElephantRunner.java
	app/com/linkedin/drelephant/util/Utils.java
	app/controllers/Application.java
	test/com/linkedin/drelephant/util/UtilsTest.java

Commit:
Add 3.sql and revert modifications in 1.sql

nntnag17 committed:
Merge pull request #109 from nntnag17/master

Aggregated Metrics Feature

paulbramsen committed:
Make flow search more powerful (#99)

paulbramsen committed:
Add thread to all log output for better debugging (#108)

paulbramsen committed:
Fix Flow/Job ID/Url mixups (#100)

stiga-huang committed:
fix a log error when throttling spark event log (#114)

rajagopr committed:
LIHADOOP-21568: Add Dropwizard metrics to Dr. Elephant. This change enables metrics by default. (#115)

Following endpoints are exposed
	1) /ping
	2) /healthcheck
	3) /metrics

Following stats are available at present
  1. Application status - /ping
  2. Queue Size
  3. Skipped Jobs Count
  4. Retry Queue Size
  5. Thread Deadlocks - /healthcheck
  6. GC Count
  7. GC Time
  8. GC Count to Uptime ratio
  9. Heap Memory Usage stats

paulbramsen committed:
Add default for boolean unboxing (#119)

paulbramsen committed:
Fix sync issue in info extractor (#118)

Fixed it by using a static initializer which I think is a good choice because

1. We don't have to add a call to initializer code anywhere (either within or outside of InfoExtractor)
2. I was able to make the instance variable final
3. Automatic thread safety is provided at the language level

paulbramsen committed:
Modify main run loop to be inline with thread pool executor paradigms (#121)

This commit brings Dr. Elephant's main producer/consumer run loop in line with standard Java best practices by submitting tasks to a work queue as Runnables making things much more robust.

Previously, if a worker thread's main while loop exited for any reason that thread would live on but would stop processing data forever since the Executor would sit waiting for additional work to be submitted. Essentially, the problem was that things were being done in a standard way and there were 2 queues. We noticed that every once in a while a worker thread's while loop would exit leading to stalled progress over time after they had all done this.

I tested this on CDH 5.3 with MapReduce 2.5.0 and Spark 1.5.2.

stiga-huang committed:
Faster fetcher for MapReduce apps (#67)

This Fetcher fetches MapReduce data from logs in HDFS directly and doesn't depend on the job history server, which is the bottleneck of the analyzing flow.

paulbramsen committed:
Show result count in search (#120)

rajagopr committed:
LIHADOOP-21657: Release Dr. Elephant v. 2.0.5 (#117)

shankar37 committed:
Add rest interface for getting resource usage by user (#125)

paulbramsen committed:
Improved metrics (#126)

* adding more metrics for monitoring process jobs count and latency

* Fix metrics null ptr API usage issues

* Add jobs processed in last 24 hours metric

akshayrai commented on app/com/linkedin/drelephant/analysis/AnalyticJobGeneratorHadoop2.java:
```diff
@@ -67,6 +69,7 @@
   private final Queue<AnalyticJob> _retryQueue = new ConcurrentLinkedQueue<AnalyticJob>();
 
   public void updateResourceManagerAddresses() {
+    String resourceManagerHttpPolicy = configuration.get(RESOURCE_MANAGER_HTTP_POLICY);
```
Can you set default to HTTP_ONLY, just in case resourceManagerHttpPolicy is null.

```String resourceManagerHttpPolicy = configuration.get(RESOURCE_MANAGER_HTTP_POLICY, HTTP_ONLY);```

akshayrai commented on app/com/linkedin/drelephant/analysis/AnalyticJobGeneratorHadoop2.java:
```diff
@@ -76,7 +79,12 @@ public void updateResourceManagerAddresses() {
-            String resourceManager = configuration.get(RESOURCE_MANAGER_ADDRESS + "." + id);
+            String resourceManager;
+            if (resourceManagerHttpPolicy.equals("HTTPS_ONLY")){
+              resourceManager = "https://" + configuration.get(RESOURCE_MANAGER_HTTPS_ADDRESS + "." + id);
```
Would prefer using final variables to substitute http:// and https://.

akshayrai commented on app/com/linkedin/drelephant/analysis/AnalyticJobGeneratorHadoop2.java:
```diff
@@ -76,7 +79,12 @@ public void updateResourceManagerAddresses() {
+            if (resourceManagerHttpPolicy.equals("HTTPS_ONLY")){
+              resourceManager = "https://" + configuration.get(RESOURCE_MANAGER_HTTPS_ADDRESS + "." + id);
+            } else {
+              resourceManager = "http://" + configuration.get(RESOURCE_MANAGER_HTTP_ADDRESS + "." + id);
```
Set a default value while retrieving configuration...here and elsewhere

probably localhost:8088

akshayrai commented:
Thanks for the PR @damienclaveau. I didn't get a chance to review it completely today but I have added a few comments. Will finish my review by tomorrow.

Adding @shankar37 @nntnag17 to review.

damienclaveau commented:
Thanks for the review.

I would like to rely more on the Hadoop codebase (org.apache.hadoop.http.HttpConfig, org.apache.hadoop.yarn.conf.YarnConfiguration.getSocketAddr, org.apache.hadoop.yarn.webapp.util.WebAppUtils.getRMWebAppURLWithScheme), 
... but there has been a lot of refactoring in theses methods between Hadoop 2.3 and Hadoop 2.7
Such dependencies will probably break backward compatibility.

Please, what is the minimal hadoop version that dr-elephant must support ?

akshayrai commented:
@damienclaveau , as far as I know, Dr. Elephant currently supports all the hadoop 2.x versions.

shankar37 committed:
Add Metrics feature for Spark Jobs (#135)

Includes resource usage, resource wastage and runtime metrics for Spark.

rajagopr committed:
Changes to load an agent jar along with Dr. Elephant. Current use-case is to publish the metrics
to other applications. The property is disabled by default and users who wish to make use of this
specify their own agent jar.

Commit:
Add rest end point for new user interface

Commit:
LIHADOOP-23923: Add rest end points for search results and search options

Commit:
LIHADOOP-23924 Create dashboard of Dr Elephant using ember

Commit:
LIHADOOP-23967: Setup JobHistory, FlowHistory and help pages for new Dr. Elephant UI

nntnag17 committed:
Merge pull request #150 from nntnag17/UIHistoryHelp

LIHADOOP-23967: Setup JobHistory, FlowHistory and help pages for new UI

Commit:
LIHADOOP-24102 Create Workflow, Job and Application pages in ember

Commit:
Added License to all the new files

Commit:
LIHADOOP-24211 Implement search page and panel

Commit:
Integration of the ember app with play framework

    Other fixes:
    * Additional information to workflow, workflow-summaries rest end point
    * Added an rest end point for the dashboard summary

nntnag17 committed:
Merge pull request #157 from nntnag17/UIHistoryHelp

Integration of the Dr. Elephant ember app with play framework

shankar37 committed:
Return the resource Usage per user as GBHours instead of MBSeconds (#158)



* Adding tests for SparkAggregated Metrics

* Fix code review comments

* Make Memory Buffer Percentage Configurable. Disable Schedule Delay for spark

* Fix code to not use getOrDefault

* use parseDouble instead for faster performance

* return resource usage by user in GB Hours

* Handle NumberFormatException when map/reduce memory config is not defined

* build fix

nntnag17 committed:
Multiple fixes on Dr. Elephant: (#159)

        * Added not-found page
	* Search panel on pages served by play
	* Piwik integration

nntnag17 committed:
Increment version to 2.0.6 (#160)

nntnag17 committed:
Multiple fixes: (#161)

LIHADOOP-24687 Dr Elephant shows more than 100% waittime for some jobs
    LIHADOOP-24688 Dr Elephant shows NAN when the resource usage and resource wasted is zero
    LIHADOOP-24695 Dr Elephant should show loading panel on search page

NoamShaish committed:
Oozie scheduler support (#152)

babak-altiscale committed:
Add null checks around update methods (#164)

Commit:
logger actually returned number of application types, not job types (#169)

* logger actually returned number of application types, not job types

* also log the appType size

Commit:
add exclude_jobs_filter to heuristics that is checked in AnalyticJob before running heuristic (#168)

Commit:
fix superParentId - now also returns the Coordinator id for subworkflows (#167)

also fixed testDepthCalculation - our test workflow only has one parent so Depth should be 1.

Commit:
use an ordered map to preserve order of jobs (#170)

rayortigas committed:
Rewrite Spark fetcher/heuristics. (#162)

* Rewrite Spark fetcher/heuristics.

The purpose of this update is to:

- rewrite the Spark data fetcher to use Spark event logs minimally, since it can be expensive to download and process these fully as done before
- rewrite the Spark data fetcher to use the [Spark monitoring REST API](https://spark.apache.org/docs/1.4.1/monitoring.html#rest-api), which provides almost all of the information Spark heuristics need
- update the Spark heuristics to provide hopefully more useful information and avoid being arbitrarily restrictive

The new Spark-related code is provided in parallel to the old Spark-related code. To enable it:

- Uncomment and swap in the appropriate fragments in `AggregatorConf.xml`, `FetcherConf.xml`, and `HeuristicConf.xml`.
- Set `SPARK_CONF_DIR` (or `SPARK_HOME`) to an appropriate location so that Dr. Elephant can find `spark-defaults.conf`.

Heuristics added:

- "Executor shuffle read bytes distribution": We now provide a distribution with min/25p/median/75p/max, with severity based on a max-to-median ratio.
- "Executor shuffle write bytes distribution": We now provide a distribution with min/25p/median/75p/max, with severity based on a max-to-median ratio.

Heuristics changed:

- "Average input size" -> "Executor input bytes distribution": Instead of providing an average along with min/max, we now provide a distribution with min/25p/median/75p/max, with severity based on a max-to-median ratio.
- "Average peak storage memory" -> "Executor storage memory used distribution": Instead of providing an average along with min/max, we now provide a distribution with min/25p/median/75p/max, with severity based on a max-to-median ratio.
- "Average runtime" -> "Executor task time distribution": Instead of providing an average along with min/max, we now provide a distribution with min/25p/median/75p/max, with severity based on a max-to-median ratio.
- "Memory utilization rate" -> "Executor storage memory utilization rate": This seemed to imply total memory but it is just the utilization rate for storage memory, so has been relabeled to indicate that. Shuffle memory is important too (but we don't seem to have access to shuffle memory utilization metrics).
- "Total memory used at peak" -> "Total executor storage memory used": This also refers to storage memory. It has been relabeled to indicate that.
- "Spark problematic stages" -> ("Spark stages with high task failure rates", "Spark stages with long average executor runtimes"): This was a combination of stages with high task failure rates and those with long runtimes. Those have been separated.

Heuristics removed:

- spark.executor.cores: I think this is somewhat discretionary. At the very least, our internal recommendation stopped matching the one in Dr. Elephant.
- spark.shuffle.manager: This was changed to "sort" by default as of Spark 1.2, so there is no current use for checking this setting.
- "Average output size": Metrics related to output size appear to be deprecated or non-existent, so there is no current use for checking this setting.

Finally, overall waste metrics are calculated based on allocation [app runtime * # of executors * executor memory] vs. usage [total executor run time * executor memory]. They were previously calculated based only on storage memory and some 50% buffer, which I didn't understand.

Added unit tests and also tested against our internal cluster as much as practically I could. Will need help to fully validate.

akshayrai committed:
LIHADOOP-24958: Investigate the cause of Skipped Jobs Alert in Dr. Elephant

akshayrai committed:
Post Refactoring of Spark Fetcher, we have removed Java 6 support

bpitman committed:
Resources used (#173)

* Fix default (mb) and verify memory size >= 0.
* Parse the string that's already been retrieved from the config.

shkhrgpt committed:
Removes dependencies on Play to load classes (#179)

akshayrai committed:
Post Refactoring of Spark Fetcher, we have removed Java 6 support (#172)

paulbramsen committed:
Add JAVA_OPTS env var export to start.sh (#116)

Commit:
LIHADOOP-25142: Build the user-summary page for Dr. Elephant (#174)

*     LIHADOOP-25142: Build the user-summary page for Dr. Elephant
      LIHADOOP-20704: Expose rest interface to provide aggregation and filtering based on username
    Other fixes:
        Tooltips on search panel
        Fix to search by job_*

shkhrgpt committed:
Removes play dependencies to read config file and read system properties. (#183)

akshayrai committed:
LIHADOOP-25809: Update Dr. Elephant README to point to Quick Setup and How to Contribute page (#185)

Quick Setup: https://github.com/linkedin/dr-elephant/wiki/Quick-Setup-Instructions
How to Contribute?: https://github.com/linkedin/dr-elephant/wiki/How-to-Contribute%3F

vaxvms committed:
Scheduler-less support for flow and job history (#181)

* Use map-reduce job configuration properties to provide flow and job
history run outside of a scheduler. They are injected into the job conf by pig (see PIG-3048) and hive (HIVE-3708)

* Prefix jobDefId/workflowDefId with username

Commit:
Show exceptions for failed workflows (#188)

rajagopr committed:
Added new heuristic DistributedCacheLimit heuristic. (#187)

Jobs which put large files(> 500MB) in the distributed cache are flagged.
Files as part of the following are considered.
  mapreduce.job.cache.files
  mapreduce.job.cache.archives

shkhrgpt committed:
Cleanes up MapReduceTaskData class by removing unnecessary constructors (#202)

shkhrgpt committed:
Fixes Spark REST fetcher for client mode applications (#193)

* Removes pattern matching

stiga-huang committed:
Fix for null pointers in TaskList returned by MapReduceFSFetcherHadoop2 (#203)

(1) Use ArrayList instead
(2) Add unit test for this

Commit:
Fix #162 with the right calculation for resourceswasted and  add missing workflow links (#207)

Commit:
Fix Exception thrown when JAVA_EXTRA_OPTIONS is not present (#210)

shkhrgpt committed:
Adds an option to fetch recently finished apps from RM (#212)

shkhrgpt committed:
Fixes issue caused by http in history server config property (#217)

stiga-huang committed:
add config for timezone of job history server (#214)

cwsteinbach commented:
Hi @damienclaveau @akshayrai, I'd like to get this committed, but it's not immediately clear to me if there's anything left to do. Are there any remaining action items or is this patch ready for a final review? Thanks.

akshayrai commented:
@damienclaveau, can you update the PR fixing all the conflicts? Do you have anything left to add to this PR?

akshayrai committed:
Include reference to the weekly meeting

shkhrgpt committed:
Fixes MapReduce aggregator and heuristic to correctly handle task data when sampling is enabled (#222)

superbobry committed:
Spark fetcher is now able to fetch event logs via REST API (#225)

This commit allows Dr. Elephant to fetch Spark logs without universal
read access to eventLog.dir on HDFS. SparkFetcher would use SparkRestClient
instead of SparkLogClient if configured as

    <params>
      <use_rest_for_eventlogs>true</use_rest_for_eventlogs>
    </params>

The default behaviour is to fetch the logs via SparkLogClient/WebHDFS.

jimdowling commented:
Is there anything happening with this PR @damienclaveau  ? We would really like to see it included.

Commit:
Update #224 (credits: rayortigas) to add FSFetcher as a standalone fetcher (#232)

Remove backup for Rest Fetcher and make Legacy FSFetcher as top level fetcher. Change the default fetcher in the config.

Commit:
Spark metrics aggregator fix (#237)

* Fix SparkMetricsAggregator to not produce negative  ResourceUsage

Commit:
Minor bug fixes in exception and UI (#238)

shkhrgpt committed:
Updates Spark configuration heuristic severity calculations (#229)

shkhrgpt committed:
Enables SparkFetcher to only get eventLog via rest and process it locally. (#243)

damienclaveau commented:
Hi everyone, with the major refactoring done to the SparkFSFetcher, this PR won't easily rebase.
I will have a look at this it in the next few days, and test everything against our clusters.

Commit:
Refactor statusapiv1 to trait and implement for ease of creation of these objects when we implement our own parser (#248)

Commit:
MRfetcher ignores failed tasks (#249)

* We have been ignoring Failed Tasks in calculation of resource usage. This handles that.
* Fixes Exception heuristic which was supposed to give the stacktrace.

ashangit committed:
Add index on severity, finish_time to speed up welcome page display (#250)

georgewu2 committed:
Add pinball scheduler to dr-elephant (#253)

georgewu2 committed:
add s3, s3a, s3n bytes read and bytes written, and update heuristics to use them (#254)

Commit:
Add custom flowtime per scheduler (#268)

Commit:
Add filtering on Job Definition Id in the Search view (#269)

skakker committed:
added logic for map reduce time-skew heuristic (#267)

ericdill committed:
DOC: Clarify the wording of output (#282)

MaximeKestemont committed:
Fix text box for ids composed of several words (#265)

akshayrai commented:
@damienclaveau , did you get a chance to rebase and test this?

jimdowling commented:
@damienclaveau  - would be great to have SSL/TLS support.

akshayrai committed:
HadoopSecurity should be a Singleton (#284)

* HadoopSecurity should be a singleton

* Update build to the last recognized successful deployment (Linkedin)

shkhrgpt committed:
BUGFIX: Fixes NullPointerException in AnalyticJobGeneratorHadoop2 (#294)

BUGFIX: Uses Long.parseLong method to fix NullPointerException issue AnalyticJobGeneratorHadoop2

skakker committed:
Fix MapReduceFSFetcherHadoop2 Fetcher filesystem to pick the configured URI (#292)

shahrukhkhan489 committed:
BUGFIX Updating AnalyticJobGeneratorHadoop2.java to resolve the Job listing Conflict (#302)

shahrukhkhan489 committed:
BUGFIX Updating java_args to elephant.conf for resolving argument conflicts (#299)

shahrukhkhan489 committed:
TUNING Updating default MR fetcher for performance (#300)

akshayrai committed:
Dr. Elephant should check for finished directory before listing

Commit:
Add httpcore dependency to solve classpath issues (#308)

damienclaveau committed:
Support for SSL
You just need to add jvm_args="-Djavax.net.ssl.trustStore=path_to_truststore" in elephant.conf

shkhrgpt commented:
Would this work when HTTPS authentication is enabled?

damienclaveau commented:
Well, this PR works well on our clusters. 
But I'm afraid the PR commit history doesn't reflect my git rebase operations.
The PR consists of only one commit actually : https://github.com/damienclaveau/dr-elephant/commit/2609d3027f4f1ee4ffee9549321c27dcd0ccae30

bannikodu commented:
esourceManager Web Application HTTPS Port (TLS/SSL) - set with port 8090, so http - 8088 doesnt open cos of SSL settings.
is it possible for Dr elephant to read using https - 8090 port.
please help on this.
